### PR TITLE
Support added for CodeType value in itemExtractionContext extension

### LINKED ISSUE #1770: Definition Based Extraction: Support code type for itemExtractionContext extension value.
**Describe the bug**
Currently during definition based extraction, the value for itemExtractionContext extension is cast as Expression. As per the [SDC FHIR IG](http://hl7.org/fhir/uv/sdc/STU3/StructureDefinition-sdc-questionnaire-itemExtractionContext.html) value can be of type valueCode.

**Component**
SDC library

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1770

**Description**
CodeType value support added in itemExtractionContext extension in ResourceMapper

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one:  Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -767,44 +767,52 @@ private fun Base.asExpectedType(): Type {
...
-private val Questionnaire.QuestionnaireItemComponent.itemExtractionContextNameToExpressionPair:
-  Pair<String, String>?
-  get() = this.extension.itemExtractionContextExtensionNameToExpressionPair
+private val Questionnaire.QuestionnaireItemComponent.itemExtractionContextExtensionValue: String?
+  get() = this.extension.itemExtractionContextExtensionValue
 
 /**
  * [Pair] of name and expression for the item extraction context extension if one and only one such
  * extension exists, or null otherwise.
  */
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:800]: this needs to be updated. and please add a link to the relevant section in the sdc ig.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
i think in the case of expression, we need to update an exsiting resource. in that case i think we need to use the interface @aditya-07 added to load up the existing resource.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -767,44 +767,53 @@ private fun Base.asExpectedType(): Type {
...
- * [Pair] of name and expression for the item extraction context extension if one and only one such
- * extension exists, or null otherwise.
+ * The item extraction context extension value of type Expression or CodeType if one and only one
+ * such extension exists, or null otherwise. See
+ * http://hl7.org/fhir/uv/sdc/STU3/StructureDefinition-sdc-questionnaire-itemExtractionContext.html
  */
-private val List<Extension>.itemExtractionContextExtensionNameToExpressionPair
+private val List<Extension>.itemExtractionContextExtensionValue
   get() =
     this.singleOrNull { it.url == ITEM_CONTEXT_EXTENSION_URL }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:804]: can you add a comment to clarify that if there are multiple item extraction context extensions, we will ignore them.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -767,44 +767,53 @@ private fun Base.asExpectedType(): Type {
...
       ?.let {
-        val expression = it.value as Expression
-        expression.name to expression.expression
+        when (it.value) {
+          is Expression -> {
+            val expression = it.value as Expression
+            expression.expression
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:809]: if you're gonna fix this in a follow up pr please add a todo and create an issue.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 3 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:811]: Created issue #1799 for follow up PR

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 4]
```diff
@@ -767,44 +767,55 @@ private fun Base.asExpectedType(): Type {
...
 /**
- * [Pair] of name and expression for the item extraction context extension if one and only one such
- * extension exists, or null otherwise.
+ * The item extraction context extension value of type Expression or CodeType if one and only one
+ * such extension exists, or null otherwise.
  */
-private val Questionnaire.itemExtractionContextExtensionNameToExpressionPair
-  get() = this.extension.itemExtractionContextExtensionNameToExpressionPair
+private val Questionnaire.itemExtractionContextExtensionValue
+  get() = this.extension.itemExtractionContextExtensionValue
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:779]: kinda feel like this function could just be inlined...

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 5]
```diff
@@ -767,44 +767,55 @@ private fun Base.asExpectedType(): Type {
...
- * [Pair] of name and expression for the item extraction context extension if one and only one such
- * extension exists, or null otherwise.
+ * The item extraction context extension value of type Expression or CodeType if one and only one
+ * such extension exists, or null otherwise.
  */
-private val Questionnaire.QuestionnaireItemComponent.itemExtractionContextNameToExpressionPair:
-  Pair<String, String>?
-  get() = this.extension.itemExtractionContextExtensionNameToExpressionPair
+private val Questionnaire.QuestionnaireItemComponent.itemExtractionContextExtensionValue: String?
+  get() = this.extension.itemExtractionContextExtensionValue
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:795]: same here

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**